xonsh/completer.py

```diff
@@ -78,16 +78,54 @@ def _normpath(p):
 
     return p
 
-class Completer(object):
-    """This provides a list of optional completions for the xonsh shell."""
+class KnownCommands(object):
 
     def __init__(self):
-        # initialize command cache
         self._path_checksum = None
         self._alias_checksum = None
         self._path_mtime = -1
         self._cmds_cache = frozenset()
+    
+    def all_commands(self):
+        """find all commands available in PATH
+        """
+        path = builtins.__xonsh_env__.get('PATH', [])
+        # did PATH change?
+        path_hash = hash(tuple(path))
+        cache_valid = path_hash == self._path_checksum
+        self._path_checksum = path_hash
+        # did aliases change?
+        al_hash = hash(tuple(sorted(builtins.aliases.keys())))
```

Comment:
> Me culpa here

Looking at git blame, I think this was me :)  Or maybe we can share the blame.

> Isn't possible for different instances of the frozenset with the same values to hash differently?

Nope. The hash of a frozenset only depends on the objects contained in it (and not on their order or the particular instance we're dealing with).  Source code for cpython's frozenset hashing function is [here](https://hg.python.org/cpython/file/tip/Objects/setobject.c#l761).


We could maybe run into trouble if someone used a mutable type as a key in the aliases dict, but that would cause this to break even if we stuck with hashing the sorted tuple.